### DISCUSSION TITLE
Support npm packages from a private GitHub repository

### DISCUSSION THREAD

**Discussion by idahogurl** [renovatebot/renovate#16345] [Request Help] [answered]
### What would you like Renovate to be able to do?

We need Renovate to support packages from GitHub just like NPM does. Here is an example. 

```
{
  "dependencies": {
   "node_project": "git+ssh://git@github.com:user_name/node_project.git@1.0.0"
   }
}
```

See https://www.pluralsight.com/guides/install-npm-packages-from-gitgithub

### If you have any ideas on how this should be implemented, please tell us here.

I'd take a look on how NPM implements the feature. I did find they use this package @npmcli/git

### Is this a feature you are interested in implementing yourself?

Yes

**Comment by rarkins** [comment, marked as answer]
Renovate already supports packages from github. But it does not support git+ssh. If you change your definition to support https, it should work already.

**Comment by idahogurl** [comment]
@rarkins Does it support authentication with https? This is a private repository. Should have mentioned that.

**Reply by rarkins** [reply]
Yes, it tries inserting the token into the URL which is used. Is it within the same org as the repo you are renovating, and are both private? And how are you running Renovate - self-hosted or using the github.com app?

**Reply by idahogurl** [reply]
Yes, it is in the same org and both are private. We are using the github.com app. I haven't tried it just yet. I just noticed in the logs that it skipped the package since it's using `git:ssh`. I'll give it a try and let you know. Would be good to get this into the documentation, if it works.

**Comment by idahogurl** [comment]
Switched to HTTPS and having the same issue as https://github.com/renovatebot/renovate/discussions/8789 

```
{
            "depType": "dependencies",
            "depName": "core-common",
            "currentValue": "https://github.com/equinox-platformx/core-common.git#semver:~2.23.1",
            "skipReason": "unversioned-reference",
            "prettyDepType": "dependency",
            "lockedVersion": null,
            "depIndex": 7,
            "updates": []
          },
```

**Reply by rarkins** [reply]
I haven't seen that semver syntax before and don't think we support it. Can you create a reproduction and file a feature request for that specifically?

**Reply by idahogurl** [reply]
I totally would but at this point I think we are going to move the package to GitHub Packages as a private package. It's the same amount of work of switching from SSH to HTTPS.

**Reply by rarkins** [reply]
You don't plan to use that semver syntax anymore or don't plan to use git dependencies anymore?

**Reply by idahogurl** [reply]
Git dependencies

**Comment by timiyay** [comment]
I've hit this same issue when experimenting with Renovate as an alternative to Dependabot.

I'm unsure if switching to `git+https` is viable for us or not. It requires a totally different auth setup to the SSH-based one that has been working for us. I can't hardcode my personal creds into a `package.json` shared by many, and used by automated CI/CD processes.

What would be involved in adding support for `git+ssh` @rarkins ? Would it be do-able for a community contribution from myself?

**Reply by rarkins** [reply]
Is the private dependency hosted on github.com?

**Reply by timiyay** [reply]
@rarkins yes it's on github.com

**Reply by rarkins** [reply]
I had thought there's a way to tell npm "either ssh or https". It seems that one allowed syntax is like `github:org/repo`. If you put that in your `package.json` instead of the `git+https` syntax, does it still work over ssh?

The challenge we have with the hosted app is that github apps do not have the concept of SSH auth. i.e. we can use the app token over https to clone sibling repos but not use SSH.

Next challenge is that we could rewrite git+ssh to https but then I wonder if your `package-lock.json` files would have https references instead of ssh?

**Reply by timiyay** [reply]
No SSH - sounds like quite a challenge!

The different syntax does work with our existing SSH-based processes. The hiccup is it's likely to create churn in our `yarn.lock` file.

When Renovate bumps the dependency, it'll rewrite the `yarn.lock` to use `git+https`. When a dev runs `yarn install` locally, it'll change it to `git+ssh`. There can be a constant churning of this file if local installs are accidently committed, which is easy to miss.

We've had this churn issue with some other dependencies in the past, and had managed to rid ourselves of it with Dependabot auto-bumps. We're trialling Renovate, and were hoping to maintain this private repo's auto-bumping.

**Reply by rarkins** [reply]
As far as I'm aware, Dependabot would also need to act like an app, meaning same non-ssh limitation,so I'm curious how that could work too. I guess potentially we could add a post update option to rewrite all git+https to git+ssh or something like that. A bit hacky but could work.

**Reply by timiyay** [reply]
I'll do some experiments with different options, and see if a path-of-least-resistance opens up.

We could solve in other ways, like publishing the private repo as a private npm package. For this specific repo, we may end up migrating its code into its single consumer repo. It was originally thought it would be consumed in multiple places, but that hasn't eventuated, so I could skip this whole problem by moving the code :)
